datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt

```diff
@@ -56,59 +59,164 @@ internal object ResourceMapper {
    * `questionnaire` and `questionnaireResponse`.
    */
   fun extract(questionnaire: Questionnaire, questionnaireResponse: QuestionnaireResponse): Base {
-    return questionnaire
-      .itemContextNameToExpressionMap
-      .values
-      .first()
-      .let { Class.forName("org.hl7.fhir.r4.model.$it").newInstance() as Base }
-      .apply { extractFields(questionnaire.item, questionnaireResponse.item) }
+    val className = questionnaire.itemContextNameToExpressionMap.values.first()
+    val base = Class.forName("org.hl7.fhir.r4.model.$className").newInstance() as Base
+    base.extractFields(questionnaire.item, questionnaireResponse.item)
+    return base
+  }
+
+  /**
+   * Extracts value for fields in the QuestionnaireResponse from the corresponding questions and
+   * answers in [questionnaireItemList] and [questionnaireResponseItemList] and updates it on the
+   * Base object target property. This method handles nested fields in the Questionnaire
+   */
+  private fun Base.extractFields(
+    questionnaireItemList: List<Questionnaire.QuestionnaireItemComponent>,
+    questionnaireResponseItemList: List<QuestionnaireResponse.QuestionnaireResponseItemComponent>
+  ) {
+    val questionnaireItemListIterator = questionnaireItemList.iterator()
+    val questionnaireResponseItemListIterator = questionnaireResponseItemList.iterator()
+    while (questionnaireItemListIterator.hasNext() &&
+      questionnaireResponseItemListIterator.hasNext()) {
+      val questionnaireItem = questionnaireItemListIterator.next()
+      val questionnaireResponseItem = questionnaireResponseItemListIterator.next()
+
+      extractField(questionnaireItem, questionnaireResponseItem)
+    }
+  }
+
+  /**
+   * Extracts value for field in the QuestionnaireResponse from the corresponding question and
+   * answer in [questionnaireItem] and [questionnaireResponseItem] and updates it on the Base object
+   * target property. This method handles nested fields in the Questionnaire
+   */
+  private fun Base.extractField(
+    questionnaireItem: Questionnaire.QuestionnaireItemComponent,
+    questionnaireResponseItem: QuestionnaireResponse.QuestionnaireResponseItemComponent
+  ) {
+    if (questionnaireItem.definition == null) {
+      if (questionnaireItem.item != null && questionnaireResponseItem.item != null) {
+        this.extractFields(questionnaireItem.item, questionnaireResponseItem.item)
+      }
+      return
+    }
+
+    val targetFieldName = questionnaireItem.definitionFieldName ?: return
+    if (targetFieldName.isEmpty()) {
+      return
+    }
+
+    val propertyType = questionnaireItem.inferPropertyResourceClass
```

Comment:
```suggestion
    val propertyType = questionnaireItem.inferPropertyResourceClass ?: return
```

and you can remove line 112 and line 133

Reply:
Fixed.